### ISSUE TITLE
Inconsistency of Container::getValues between php 7.1 and 7.2

### ISSUE THREAD

**Issue by slepic** [nette/forms#232]
Version: 3.0.1

### Bug Description
If you add components to a Container and their names are integer-like, calling getValues() on the container will yield different result in PHP <= 7.1 then it will in PHP >= 7.2.

Reason is this conversion of numeric keys in array object conversion
https://www.php.net/manual/en/migration72.incompatible.php
and this piece of code:
https://github.com/nette/forms/blob/v3.0.1/src/Forms/Container.php#L117

where it builds the values hash from keys of array of components rather then the components' names. key of the array is implicitly converted to int.
Since PHP <= 7.1 will convert the int key to string again once it used as property name,
but PHP >= 7.2 will allow to have an object with integer named property.
Comparisions of the property name with type strict coparision operator `===` then yield different results.

### Steps To Reproduce
```
$form = new Form();
$form->addText("1");

$values = $form->getValues();
foreach ($values as $key => $value) {
   gettype($key); //7.1: string; 7.2: int
}
```

### Expected Behavior
assert(is_string($key))

### Possible Solution
cast $name to string before using it as property name on lines 119 and 121
https://github.com/nette/forms/blob/v3.0.1/src/Forms/Container.php#L117

**Comment by dg** [comment]
As you said, the reason is changed behavior in PHP 7.2 vs 7.1. It is a PHP engine matter and cannot be changed by the framework.

**Comment by slepic** [comment]
I couldn't agree less. Any package that claims to support both PHP 7.1 and 7.2 should have consistent behaviour across those supported versions. The code was obviously written in times when <= 7.2 was about.  Then when 7.2 came out. It should have either been said that 7.2 is not yet supported or made sure everything works the same in 7.2. This didn't happen. Now anyone who is using nette and migrates from 7.1 to 7.2 and follows PHP's migration guide to the last point is still potentialy subject to BC breaks even tho he handled all of his code. Just because he cannot handle 3rd party code that didnt take the same effort to migrate to support 7.2. All the components have string names, and there is no reason why the names should change to integers when upgrading php version.
So to sum up, this IS an issue of nette framework. And nette SHOULD fix it. Even though this is just a minor break and probably won't affect many. In principle, it is nettte's responsibility to fix it. And I am SERIOUSLY disappointed, that you dont see it.

**Comment by dg** [comment]
If you know how to fix it, write me. I have no idea.

**Comment by slepic** [comment]
Just update this:
```
foreach ($this->getComponents() as $name => $control) {
  if ($control instanceof IControl && !$control->isOmitted()) {
    $obj->$name = $control->getValue();
  } elseif ($control instanceof self) {
    $obj->$name = $control->getValues($isArray ? self::ARRAY : null);
  }
}
```
to this:
```
foreach ($this->getComponents() as $name => $control) {
  $name = (string) $name;
  if ($control instanceof IControl && !$control->isOmitted()) {
    $obj->$name = $control->getValue();
  } elseif ($control instanceof self) {
    $obj->$name = $control->getValues($isArray ? self::ARRAY : null);
  }
}
```
https://github.com/nette/forms/blob/v3.0.1/src/Forms/Container.php#L117

**Comment by dg** [comment]
Did you test it?

**Comment by slepic** [comment]
Well I didn't run tests of the nette forms repo. But I ran tests of a dependent package which actualy got broken on this and thats how i found out... When I made sure the values hash keys are cast to strings the tests fixed. And this is the same thing, except it is done yet in the nette core. And anyway logicaly there is nothing that can happen other then making the 7.1/7.2 difference fade away as 7.1 is doing this conversion for you automaticaly...

**Comment by slepic** [comment]
Or maybe this should be handled by `RecursiveComponentIterator::key()` already. The component names are always string, even if contain just a number. Then it makes sense that whenever components are iterated through `getComponents()`, the iterators must return string 
```
public function key(): string {
  return (string) parent::key();
}
```

**Comment by dg** [comment]
I am not talking about nette repo tests, I am just asking if your suggested fix has solved your problem?

**Comment by slepic** [comment]
Aha it does not. lol. I can finaly see it. Sry to have wasted your time... :man_facepalming:
